rfcs/0001-subgraph-composition.md

```diff
@@ -0,0 +1,233 @@
+}
+```
+
+This makes queries like the following possible, where you can go "back" from
+accounts and contracts to proposal entities, despite the Ethereum `Address` type
+originally being defined in the `ethereum/mainnet` subgraph.
+
+```graphql
+{
+  accounts {
+    address
+    proposals {
+      id
+      proposer {
+        address
+    }
+  }
+}
+```
+
+In the above case, the `proposals` field on the extended type is derived, which
+means that an implementation likely wouldn't have to create a local extension
+type in the store. However, if `proposals` was defined as
+
+```graphql
+extend type Address {
+  proposals: [Proposal!]!
+}
+```
+
+then it would have to be possible for the subgraph mappings to create `Address`
+instances and set the `proposals` field. These entity instances would have to be
+merged with the corresponding original entities from the `ethereum/mainnet`
+subgraph.
+
+### Subgraph availability
+
+In the decentralized network, queries will be split and routed through the
+network based on what indexers are available and which subgraphs they index.
+At that point, failure to find an indexer for a subgraph that types were
+imported from will result in a query error.
+
+Until the network is reality, we are dealing with individual Graph nodes and
+querying subgraphs where foreign entities are not also indexed on the same node
+should be handled with more tolerance. This RFC proposes that entity references
+to a subgraph that is not available should simply resolve to `null` values in
+query results.
+
+## Compatibility
+
+Subgraph composition is a purely additive, non-breaking change. Existing
+subgraphs remain valid without any migrations being necessary.
+
+## Drawbacks and Risks
+
+Reasons that could speak against implementing this feature:
+
+- Schema parsing and validation becomes more complicated. Especially validation
+  of foreign schemas may not always be possible, depending on whether and when
+  the referenced subgraph is available on the Graph node or not.
+
+- Query execution becomes more complicated. The subgraph a type belongs to must
+  be identified and local and foreign versions of extended entities have to be
+  queried separately and be merged.
+
+## Alternatives
+
+No alternatives have been considered.
+
+There are other ways to compose subgraph schemas using GraphQL technologies such
+as [schema
+stitching](https://www.apollographql.com/docs/graphql-tools/schema-stitching/)
+or [Apollo
+Federation](https://www.apollographql.com/docs/apollo-server/federation/introduction/).
+However, schema stitching is being deprecated and Apollo Federation requires a
+centralized server to serve a combined GraphQL API.
+
+Another reason not to use these is that GraphQL will only be _one_ of several
+query languages supported in the future. Composition therefor has to be
+implemented in a query-language-agnostic way.
+
+## Open Questions
+
+- An open question is whether it should be possible to extend foreign
```

Comment:
What about implementing interfaces for foreign types?

Reply:
It's going to fun to mix local and foreign concrete types that implement the same interface (local or foreign). It sounds feasible though.

Reply:
So this is more of a future feature than an open question to be tackled in this RFC?

Reply:
I think we should tackle it. Will think about this a bit more.

Reply:
Done, added a section about interfaces.